RE2022-240: Add a string filter representation class

MrCreosote opened the pull request:
Represents a filter for a string column and translates to ArangoSearch AQL lines.

Xiangs18 commented on test/src/service/data_products/filtering_test.py:
```diff
@@ -68,37 +77,79 @@ def test_to_arangosearch_aql():
             ColumnType.INT,
             " [  \t   ",
-            "^Invalid range specification; expected exactly one comma:  \\[  \t   $"
+            "Invalid range specification; expected exactly one comma:  \\[  \t   "
```
There are a bunch of  `\\` inside test_rangefilter_from_string_fail function. Why are they?

MrCreosote replied:
Regex escape sequences

Xiangs18 replied:
Oh, I see. But why not using "\"? You also want to match a single "\" in your value as well?

MrCreosote replied:
Python complains if you escape an invalid escape character:

```
  /home/crushingismybusiness/github/kbase/collections/test/src/service/data_products/filtering_test.py:89: DeprecationWarning: invalid escape sequence '\['
```

Xiangs18 commented on src/service/data_products/filtering.py:
```diff
@@ -3,26 +3,69 @@
 
 
 class SearchQueryPart(BaseModel):
+    variable_assignments: Annotated[dict[str, str], Field(
```
should be Annotated[dict[str, str] | `None`, Field(...)] = None?

MrCreosote replied:
Fixed

Xiangs18 commented:
Somehow I cannot left a comment there, but [this](https://github.com/kbase/collections/pull/451/files#diff-9fbc55937877d77ed5cb163f16993b7fdaef56d0757c8f3dff12b22c82e141f8L76) looks like a @staticmethod rather than @classmethod

MrCreosote commented:
> change _parse_inclusivity(self...) -> _parse_inclusivity(cls...)

I think it originally was a regular method, but when I switched it to a classmethod I forgot to switch the self var. Anyway, fixed

Xiangs18 commented on src/service/data_products/filtering.py:
```diff
@@ -3,26 +3,69 @@
+        strategy - the strategy for the filter.
+        string - the string to parse to create the filter.
+        """
+        raise NotImplementedError()
```
Ah, not sure about this line. Why we need to raise NotImplementedError()? 

I thought a TypeError will be automatically raised if any of abstract methods is not implemented.

MrCreosote replied:
Just to make it clear to the subclass implementer that there's no need to call the super method

Xiangs18 reviewed (approved):
LGTM!
